ramips: add support for TP-Link Archer C5 v4

luizluca opened the pull request:
This PR is based on #2174 from @dengqf6. It avoids touching existing switch drivers by
reusing the an already merged driver (from target mediatek), moving it to generic.
In mediatek, there were two copies of the driver (5.4 and 5.10) while in generic the small difference
is dealt using macros. There were some changes to enable it for ramips target and create a kmod package
only for ramips as mediatek/mt7622 uses the driver builtin.

I kept the move from mediatek to generic in small commits during the review. However, they might be squashed
into one before merge.

The router is working as expected. The only noticeable drawback is that the new driver is way bigger
than the two removed modules. Does it support more features?

- 8214    kmod-switch-rtl8366-smi_5.4.105-1_mipsel_24kc.ipk
- 9407    kmod-switch-rtl8367b_5.4.105-1_mipsel_24kc.ipk
- 97114   kmod-switch-rtl8367s_5.4.105-1_mipsel_24kc.ipk

20210322 update: change the driver approach from "copy from mediatek" to "move to generic". The error shown in dmesg is gone after fixes in DTS.
20210405 update: kernel 5.10 seems to work as expected

luizluca committed:
ramips: add support for TP-Link Archer C5 v4

TP-Link Archer C5 v4 is a dual band router with 5 GbE ports
Advertised as AC1200 for its 867Mbps (2x2) 5GHz band
and 300 Mbps (2x2) 2.4GHz band.

Specs:
- SoC: MediaTek MT7620A
- Ethernet: 5x GbE ports (Realtek RTL8367S)
- Wireless 2.4GHz: MediaTek MT7620A
- Wireless 5GHz: MediaTek MT7612E
- RAM: 64MiB
- ROM: 8MiB (GD25Q64CSIG)
- 1 USB 2.0 port
- 2 Buttons (WPS and reset)
- 8 LEDs

2.4GHz has a poor performance when compared to stock firmware, not
passing about 20 Mbps in AP mode. It uses rt2x00 driver, which still
misses a bunch of features.

Flash instructions:

Install using tftp recovery procedure:

1. Rename the openwrt-ramips-mt7620-tplink_archer-c5-v4-squashfs-tftp-recovery.bin to tp_recovery.bin
2. start a TFTP server from IP address 192.168.0.66 and serve the image tp_recovery.bin
3. connect your device to the LAN port
4. power up the router while pressing the reset button until WDS LED
   turns on

This procedure can be used to return to factory image. However images
from TP-Link has an extra 512 bytes header that needs to be removed:

 dd if=Archer_C5\(BRWISP\)v4_3.16.0_0.9.1_up_boot\(190404\)_2019-04-04_09.56.27.bin \
   of=tp_recovery.bin bs=512 skip=1

Install using serial console:

1. Rename the factory.bin to to test.bin
2. start a TFTP server from IP address 192.168.0.225 and serve the image
   named test.bin
3. connect your device to the LAN port
4. power up the router and press 4 on the console to stop the boot process.
5. enter the following commands on the router console

  tftp 0x80060000 test.bin
  erase tplink 0x20000 0x7a0000
  cp.b 0x80060000 0x20000 0x7a0000
  reset

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

dengqf6 commented on target/linux/ramips/mt7620/base-files/etc/board.d/02_network:
```diff
@@ -207,6 +207,12 @@ ramips_setup_interfaces()
+		ucidef_add_switch "switch0"
+		ucidef_add_switch_attr "switch0" "enable" "0"
+		ucidef_add_switch "switch1" \
+			"0:lan:4" "1:lan:3" "2:lan:2" "3:lan:1" "4:wan" "5@eth0"
```
Should be `7@eth0`

luizluca replied:
Not for the new driver. CPU is at port 5. Port 7 is missing.

dengqf6 replied:
Does the luci switch page show CPU port link status correctly?

luizluca replied:
![Screenshot_20210322_020202](https://user-images.githubusercontent.com/836788/111942807-e07c8200-8ab2-11eb-92b1-516a1181dc75.png)

I looks OK, although CPU port has no link. I get similar info from swconfig:

```
# swconfig dev switch1 port 5 show
...

I don't know what the old driver returned. Anyway all labels match the corresponding case label.
ifInBroadcastPkts                   : 26
ifOutDiscards                       : 0

        pvid: 1
        link: ???
```

dengqf6 commented on target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:
```diff
@@ -0,0 +1,112 @@
+	mdio0: mdio-bus {
+		status = "okay";
+		reset-gpios = <&gpio2 20 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <10000>;
```
These 2 `reset` lines should be removed.

luizluca replied:
That fixed the error in dmesg

dengqf6 commented:
> I don't know if I should simply copy the driver from mediatek or move it to generic.

I think it's better to move it to generic to avoid duplicates.

> Anyway, the router is working as expected. The only noticiable drawback is that the new driver is way bigger
> than the two removed modules. Does it support more features?

Yes, but most are unnecessary for swconfig, such as PTP, Link Aggregation, ACL, CPU tag.
Ideally there should be a DSA driver for RTL8367x series, with all those features supported.

adschm commented on target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:
```diff
@@ -0,0 +1,112 @@
+
+	rtkgsw: rtkgsw@0 {
+		compatible = "mediatek,rtk-gsw";
+		//mediatek,ethsys = <&ethsys>;
```
So?

luizluca replied:
I removed it.

adschm commented on target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:
```diff
@@ -0,0 +1,112 @@
+		//mediatek,ethsys = <&ethsys>;
+		mediatek,mdio = <&mdio0>;
+		mediatek,reset-pin = <&gpio2 20 GPIO_ACTIVE_HIGH>;
+                //status = "okay";
```
This should be dropped (okay is default anyway).

adschm commented on target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:
```diff
@@ -0,0 +1,112 @@
+	mtd-mac-address = <&rom 0xf100>;
+};
+
+
```
drop second empty line.

adschm commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -1005,6 +1005,22 @@ define Device/tplink_archer-c2-v1
+  IMAGE_SIZE := 7808k
+  TPLINK_FLASHLAYOUT := 8Mmtk
+  TPLINK_HWID := 0x04DA857C
+  TPLINK_HWREV := 0x0C000600
```
Lower case for hex letters is preferred here.

adschm commented on target/linux/ramips/modules.mk:
```diff
@@ -140,7 +140,7 @@ define KernelPackage/switch-rtl8367s
   KCONFIG:=CONFIG_RTL8367S_GSW
   FILES:=$(LINUX_DIR)/drivers/net/phy/rtk/rtl8367s_gsw.ko
-  AUTOLOAD:=$(call AutoLoad,43,rtl8367s,1)
+  AUTOLOAD:=$(call AutoLoad,43,rtl8367s_gsw,1)
```
Looks like this should be a separate commit.

adschm commented on target/linux/ramips/mt7620/base-files/etc/board.d/01_leds:
```diff
@@ -183,6 +183,10 @@ tplink,archer-c20i)
 	;;
+tplink,archer-c5-v4)
+	ucidef_set_led_switch "lan" "lan" "$boardname:green:lan" "switch1" "0x0f"
+	ucidef_set_led_switch "wan" "wan" "$boardname:green:wan" "switch1" "0x10"
```
boardname needs to be removed here (it also is in DTS as it appears).

adschm commented on target/linux/ramips/mt7620/base-files/etc/board.d/02_network:
```diff
@@ -207,6 +207,12 @@ ramips_setup_interfaces()
 		;;
+	tplink,archer-c5-v4)
+		ucidef_add_switch "switch0"
+		ucidef_add_switch_attr "switch0" "enable" "0"
```
Above it's "false" instead of "0". No idea whether that's relevant.

adschm commented:
> > I don't know if I should simply copy the driver from mediatek or move it to generic.
> 
> I think it's better to move it to generic to avoid duplicates.

Without being familiar with the details, from a quick look I'd prefer that, too.

luizluca committed:
mediatek: move rtl8367c driver to generic

It is also used by ramips/mt7620

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca committed:
generic: adapt rtl8367s for both 5.4 and 5.10

All driver files are identical between these two versions but
drivers/net/phy/rtk/rtl8367s_dbg.c.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca committed:
generic: fix patches for rtl8367s

Mention that it is also used by mt7620.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca committed:
kernel: add kmod-switch-rtl8367s for module rtl8367s_gsw

This kmod depends on ramips target as mediatek uses the driver
only as builtin.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca commented:
I just forced-pushed replacing the copied driver with a "moved from mediatek to generic" one. I added patches for 5.10 but I didn't test them in mt7620 (Is that target ready for 5.10?). I built a mt7622 target, which already uses 5.10, and I could see it built the driver correctly. However, I do not have a mt7622 for runtime tests.

I guess the first four commits could be squashed into the first one but I left them isolated to highlight each changed part during review.

luizluca commented:
I just noticed that pci wlan (5Ghz) mac was '00:00:00:00:00:b8'. After I started to use mt7620a_tplink_archer.dtsi, the old wifi definition was in conflict. I just forced-pushed a fix.

I noticed that other "archer"s use `mtd-mac-address-increment = <(-1)>;` or `<(-2)>` instead of `<2>;` (from original PR). This is a quick survey:

```
mt7620a_tplink_archer-c20i.dts: mtd-mac-address-increment = <(-1)>;
mt7620a_tplink_archer-c20-v1.dts:       mtd-mac-address-increment = <(-2)>;
mt7620a_tplink_archer-c20-v1.dts:       mtd-mac-address-increment = <(-1)>;
mt7620a_tplink_archer-c2-v1.dts:                mtd-mac-address-increment = <(-1)>;
mt7620a_tplink_archer-c50-v1.dts:       mtd-mac-address-increment = <(-2)>;
mt7620a_tplink_archer-c50-v1.dts:       mtd-mac-address-increment = <(-1)>;
mt7620a_tplink_archer-c5-v4.dts:        mtd-mac-address-increment = <2>;
```

I'm still keeping `<2>`. Is it the correct one?

Should I try ML instead of github? This PR does not have the scaring switch changes. If that switch change is also accepted in the future, we could simply change the package selection (with a minor DTS change).

frankis5 commented:
> 
> 
> I just noticed that pci wlan (5Ghz) mac was '00:00:00:00:00:b8'. After I started to use mt7620a_tplink_archer.dtsi, the old wifi definition was in conflict. I just forced-pushed a fix.
> 
> I noticed that other "archer"s use `mtd-mac-address-increment = <(-1)>;` or `<(-2)>` instead of `<2>;` (from original PR). This is a quick survey:
> 
> ```
> mt7620a_tplink_archer-c20i.dts: mtd-mac-address-increment = <(-1)>;
> mt7620a_tplink_archer-c20-v1.dts:       mtd-mac-address-increment = <(-2)>;
> mt7620a_tplink_archer-c20-v1.dts:       mtd-mac-address-increment = <(-1)>;
> mt7620a_tplink_archer-c2-v1.dts:                mtd-mac-address-increment = <(-1)>;
> mt7620a_tplink_archer-c50-v1.dts:       mtd-mac-address-increment = <(-2)>;
> mt7620a_tplink_archer-c50-v1.dts:       mtd-mac-address-increment = <(-1)>;
> mt7620a_tplink_archer-c5-v4.dts:        mtd-mac-address-increment = <2>;
> ```
> 
> I'm still keeping `<2>`. Is it the correct one?
> 

Generally it's preferred that you use whatever the OEM MAC assignments are.  For the TP-link routers I've had, 5G is usually -1 with the 2.4g using the same MAC as the LAN.  The TP-link emulators on their website using -1 for 2.4g and -2 for 5g so there's a small discrepency on that part.  +2 is used by other manufactures.

luizluca commented:
> Generally it's preferred that you use whatever the OEM MAC assignments are. For the TP-link routers I've had, 5G is usually -1 with the 2.4g using the same MAC as the LAN. The TP-link emulators on their website using -1 for 2.4g and -2 for 5g so there's a small discrepency on that part. +2 is used by other manufactures.

@dengqf6 , as the original author, can you clarify if +2 matches the OEM behavior or do I need to revert to factory?

dengqf6 commented:
> @dengqf6 , as the original author, can you clarify if +2 matches the OEM behavior or do I need to revert to factory?

Yes, it was +2 when I tested the stock firmware. I don't know if a later hw version has changed that, though.

luizluca commented:
I gave 5.10 a try enabling testing kernel and rebasing this PR over master (r16423+5-bcdf600fc5). It does work as before, including the poor 2.5Ghz performance. The only noticeable difference were with these dmesg msgs:

```
[    0.106286] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.115313] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.123205] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.129212] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.138756] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.147757] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.155633] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.161639] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.171210] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.180211] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.188087] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.194094] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.203666] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.212669] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.220545] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.226552] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.236120] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.245121] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.252994] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.259000] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.268568] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.277567] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.285442] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.291450] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.300584] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.309983] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.318984] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.326863] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.332869] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.341997] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.350276] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.359274] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.367148] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.373154] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.382282] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.388704] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.397702] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.405572] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.411578] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.420705] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.429978] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.438974] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.446849] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.452856] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.462581] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.471585] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.479459] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.485465] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.495228] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.504233] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.512108] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.518116] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
[    0.527832] OF: /palmbus@10000000/gpio@600: could not get #gpio-cells for /ethernet@10100000/mdio-bus
[    0.536832] OF: /palmbus@10000000/gpio@638: could not get #gpio-cells for /pinctrl/sdhci
[    0.544708] OF: /palmbus@10000000/gpio@660: could not find phandle
[    0.550716] OF: /palmbus@10000000/gpio@688: could not get #gpio-cells for /palmbus@10000000/intc@200
```

They do not seem to be fatal and they are not specific to C5v4 model. Anyway, the test was focused on testing the driver for 5.10, which worked as expected.

luizluca commented:
@blogic , could you please take a look on the "move rtl8367c driver to generic" and related patches?
I does not have a mediatek device to test if this could break how the driver is currently being used.

luizluca commented:
> > @dengqf6 , as the original author, can you clarify if +2 matches the OEM behavior or do I need to revert to factory?
> 
> Yes, it was +2 when I tested the stock firmware. I don't know if a later hw version has changed that, though.

Thanks! Then 2 it is.

From my side, I don't have any more pending issues. The test for mediatek target is an extra as the source code on build_dir is the same as before, and it got compiled.

Should I squash the first commits into a single "move to generic"?

adschm commented:
> Should I squash the first commits into a single "move to generic"?

Personally, I think these are separate changes, so they should stay separate.

> 
> Should I mention that 2.5Ghz wifi performance is poor in commit message?

Additional information is always valuable. Somebody observing that on his device would then be able to find out that it's "expected". So, IMO, yes, please.

luizluca commented:
> > Should I squash the first commits into a single "move to generic"?
> 
> Personally, I think these are separate changes, so they should stay separate.
> 
> > Should I mention that 2.5Ghz wifi performance is poor in commit message?
> 
> Additional information is always valuable. Somebody observing that on his device would then be able to find out that it's "expected". So, IMO, yes, please.

Debased, reworded and retested. If I should not squash, it's ready to merge.

arinc9 commented:
What's the status of this PR? I'm working on Asus RT-AC88U with rtl8365mb switch which is pretty similar to rtl8367s. It would be nice to have the kmod ready to choose from menuconfig on the trunk.

luizluca commented:
> What's the status of this PR? I'm working on Asus RT-AC88U with rtl8365mb switch which is pretty similar to rtl8367s. It would be nice to have the kmod ready to choose from menuconfig on the trunk.

It is stalled until someone with commit access takes a look. I've been using this for months.
I just hope you copied my commit related to kernel module for #4178. It would be a shame all that rework. Anyway, it would be worth it if it gets merged.

arinc9 commented:
This commit? e0219f619553a986c21cd747b4f471d7996f4919

luizluca commented:
> This commit? [e0219f6](https://github.com/openwrt/openwrt/commit/e0219f619553a986c21cd747b4f471d7996f4919)

Yes. That one and the previous ones. I only saw afterwards you mentioned this PR in commit message.

I thought about doing a standalone PR before but I was already with little hope after @dengqf6 , me and others spend years keeping out of tree patches and trying to get c5v4 support merged. This PR was probably my final try. New devices requiring this driver might bring more attention to it.

If the PR gets stalled for a couple of weeks, you could try sending it to the maillist.

luizluca commented:
Closing as #4327 seems to be getting more stream.

luizluca commented:
See #14418
